tests/test_dir_watcher.py

```diff
@@ -0,0 +1,268 @@
+    dir_watcher._on_file_created(Mock(src_path=str(f)))
+    assert file_pusher.file_changed.called == expect_called
+
+
+def test_dirwatcher_finish_uploads_unheardof_files(
+    tempdir: Path, file_pusher: FilePusher, dir_watcher: DirWatcher
+):
+    f = tempdir / "my-file.txt"
+    write_with_mtime(f, b"content", mtime=0)
+    dir_watcher.finish()
+    file_pusher.file_changed.assert_called_once_with("my-file.txt", str(f), copy=False)
+
+
+def test_dirwatcher_finish_skips_now_files(
+    tempdir: Path, file_pusher: FilePusher, dir_watcher: DirWatcher
+):
+    f = tempdir / "my-file.txt"
+    dir_watcher.update_policy(str(f), "now")
+    write_with_mtime(f, b"content", mtime=0)
+    dir_watcher.finish()
+    file_pusher.file_changed.assert_not_called()
+
+
+def test_dirwatcher_finish_uploads_end_files(
+    tempdir: Path, file_pusher: FilePusher, dir_watcher: DirWatcher
+):
+    f = tempdir / "my-file.txt"
+    write_with_mtime(f, b"content", mtime=0)
+    dir_watcher.update_policy(str(f), "end")
+    dir_watcher.finish()
+    file_pusher.file_changed.assert_called_once_with("my-file.txt", str(f), copy=False)
+
+
+@pytest.mark.parametrize("changed", [True, False])
+def test_dirwatcher_finish_uploads_live_files_iff_changed(
+    tempdir: Path,
+    file_pusher: FilePusher,
+    dir_watcher: DirWatcher,
+    changed: bool,
+):
+    f = tempdir / "my-file.txt"
+    write_with_mtime(f, b"content", mtime=0)
+    dir_watcher.update_policy(str(f), "live")
+    if changed:
+        write_with_mtime(f, b"new content", mtime=1)
+
+    file_pusher.file_changed.reset_mock()
+    dir_watcher.finish()
+    assert file_pusher.file_changed.called == changed
+
+
+@pytest.mark.parametrize("ignore", [True, False])
+def test_dirwatcher_finish_skips_ignoreglob_files(
+    tempdir: Path,
+    file_pusher: FilePusher,
+    dir_watcher: DirWatcher,
+    settings,
+    ignore: bool,
+):
+    if ignore:
+        settings.ignore_globs = ["*.txt"]
+
+    f = tempdir / "my-file.txt"
+    write_with_mtime(f, b"content", mtime=0)
+    dir_watcher.update_policy(str(f), "end")
+    dir_watcher.finish()
+    assert file_pusher.file_changed.called == (not ignore)
+
+
+@pytest.mark.skip(reason="Live *should* take precedence over Now, I think, but I don't want to change the existing behavior yet")
+def test_dirwatcher_prefers_live_policy_when_multiple_rules_match_file(
+    tempdir: Path, dir_watcher: DirWatcher
+):
+    f = tempdir / "my-file.txt"
+    write_with_mtime(f, b"content", mtime=0)
+    dir_watcher.update_policy("*.txt", "live")
+    dir_watcher.update_policy("my-file.*", "end")
+    dir_watcher.update_policy("my-*.txt", "now")
+    assert isinstance(
+        dir_watcher._get_file_event_handler(str(f), "my-file.txt"), PolicyLive
+    )
+
+
+@pytest.mark.skip(
+    reason="Surprisingly, this test fails. Do we want to change behavior to make it pass? TODO(spencerpearson)"
+)
+def test_dirwatcher_can_overwrite_policy_for_file(
```

Comment:
that's interesting, so basically in this case it keeps the original behavior?

Reply:
It does! Which policy wins is determined by the dict iteration order [here](https://github.com/wandb/client/blob/afc6d9c2bf1f80d9aed8eb6ebd47a804b324bc4d/wandb/filesync/dir_watcher.py#L294), which, at least in the later versions of Python, is determined by the dict initialization order [here](https://github.com/wandb/client/blob/afc6d9c2bf1f80d9aed8eb6ebd47a804b324bc4d/wandb/filesync/dir_watcher.py#L186-L188). (In 3.6 I think the iteration order is undefined.)

Reply:
Ah... so I like your idea to make it deterministic... since we will have an expected behavior every time...